src/proxy/resolve.rs

```diff
@@ -107,61 +125,440 @@ where
-                    let svc = self.make.make(&target);
-                    self.is_empty.store(false, Ordering::Release);
-                    return Ok(Async::Ready(Change::Insert(addr, svc)));
+                    // Start building the service and continue.
+                    // Remove any pending additions for this addr so that they
+                    // cannot race.
+                    self.make_futures.remove(&addr);
+                    let fut = self.make.call(target);
+                    self.make_futures.push(addr, fut);
+                }
+                Update::Remove(addr) => {
+                    // If the service is still pending, cancel it. It won't
+                    // actually be removed until a subsequent poll, however.
+                    // If it was canceled, we send the remove anyway, since this
+                    // may have been an update to a pre-existing addr.
+                    self.make_futures.remove(&addr);
+                    return Ok(Async::Ready(Change::Remove(addr)));
                 }
-                Update::Remove(addr) => return Ok(Async::Ready(Change::Remove(addr))),
                 Update::NoEndpoints => {
+                    // Mark the service as explicitly empty. It's expected that
+                    // Remove events have already or will be receieved to
+                    // actually empty the receiver.
                     self.is_empty.store(true, Ordering::Release);
-                    // Keep polling as we should now start to see removals.
-                    continue;
                 }
             }
         }
     }
 }
 
+impl<R, M> tower_discover::Discover for Discover<R, M>
+where
+    R: Resolution,
+    R::Endpoint: fmt::Debug,
+    R::Error: Into<Error>,
+    M: svc::Service<R::Endpoint>,
+    M::Error: Into<Error>,
+{
+    type Key = SocketAddr;
+    type Service = M::Response;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
+        if let Async::Ready(change) = self.poll_resolution()? {
+            return Ok(Async::Ready(change));
+        }
+
+        if let Async::Ready(Some((addr, svc))) = self.make_futures.poll().map_err(Into::into)? {
+            self.is_empty.store(false, Ordering::Release);
+            return Ok(Async::Ready(Change::Insert(addr, svc)));
+        }
+
+        Ok(Async::NotReady)
+    }
+}
+
+// === impl EndpointStatus ===
+
 impl EndpointStatus {
     pub fn is_empty(&self) -> bool {
         self.0.load(Ordering::Acquire)
     }
 }
+
+// === impl MakeFutures MakeFutures===
+
+impl<F: Future> MakeFutures<F> {
+    fn new() -> Self {
+        Self {
+            futures: FuturesUnordered::new(),
+            cancelations: IndexMap::new(),
+        }
+    }
+
+    fn push(&mut self, addr: SocketAddr, inner: F) {
+        let (cancel, canceled) = oneshot::channel();
+        self.cancelations.insert(addr, cancel);
```

Comment:
Also, in re: my earlier comment about moving the removal of a previous pending future to `MakeFutures::push`, it also saves us a hashmap lookup, since `IndexMap::insert` returns the prior value if there was one. We could just say
```rust
if let Some(cancel_prior) = self.cancelations.insert(addr, cancel) {
    let _ = cancel_prior.send(());
}
```

This is not too hot so that kind of optimization may not matter a whole lot, but it seems neater overall to me...